src/main/java/ch/njol/skript/structures/StructFunction.java

```diff
@@ -35,49 +35,62 @@
 import org.skriptlang.skript.lang.structure.Structure;
 
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.MatchResult;
 
 @Name("Function")
 @Description({
 	"Functions are structures that can be executed with arguments/parameters to run code.",
-	"They can also return a value to the trigger that is executing the function."
+	"They can also return a value to the trigger that is executing the function.",
+	"Note that local functions always come after global functions, in parsing and execution"
 })
 @Examples({
 	"function sayMessage(message: text):",
 	"\tbroadcast {_message} # our message argument is available in '{_message}'",
-	"function giveApple(amount: number) :: item:",
-	"\treturn {_amount} of apple"
+	"local function giveApple(amount: number) :: item:",
+	"\treturn {_amount} of apple",
 })
-@Since("2.2")
+@Since("2.2, INSERT VERSION (local functions)")
 public class StructFunction extends Structure {
 
-	public static final Priority PRIORITY = new Priority(400);
+	public static final Priority GLOBAL_PRIORITY = new Priority(400);
+	public static final Priority LOCAL_PRIORITY = new Priority(450);
 
 	private static final AtomicBoolean validateFunctions = new AtomicBoolean();
 
+	@SuppressWarnings("null")
+	private final static String functionPattern = "(" + Functions.functionNamePattern + ")\\((.*)\\)(?:\\s*::\\s*(.+))?";
+
 	static {
-		Skript.registerStructure(StructFunction.class, "function <.+>");
+		Skript.registerStructure(StructFunction.class, "[:local] function <" + functionPattern + ">");
 	}
 
 	@Nullable
 	private Signature<?> signature;
+	private boolean local;
+	private MatchResult regex;
 
 	@Override
 	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult, EntryContainer entryContainer) {
+		local = parseResult.hasTag("local");
+		regex = parseResult.regexes.get(0);
 		return true;
 	}
 
 	@Override
 	public boolean preLoad() {
-		signature = Functions.loadSignature(getParser().getCurrentScript().getConfig().getFileName(), getEntryContainer().getSource());
-		return true;
+		signature = Functions.loadSignature(getParser().getCurrentScript().getConfig().getFileName(), local, regex);
+		return signature != null;
 	}
 
 	@Override
 	public boolean load() {
 		ParserInstance parser = getParser();
-		parser.setCurrentEvent("function", FunctionEvent.class);
+		parser.setCurrentEvent((local ? "local " : "") + "function", FunctionEvent.class);
+
+		if (signature == null) // Signature parsing failed, probably: null signature
```

Comment:
You don't need this check as load() will not be called if preLoad() returns false

You could do an assertion instead.